Add _connector API

jedrazb opened the pull request:
<!--

Hello there!

Thank you for opening a pull request!
Please make sure to follow the steps below when opening a pr:

- Sign the CLA https://www.elastic.co/contributor-agreement/
- Tag the relative issue (if any) and give a brief explanation on what your changes are doing
- If you did a spec change, remember to generate again the outputs, you can do it by running `make contrib`
- Add the appropriate backport labels. If you need to backport a breaking change (e.g. changing the structure of a type or changing the type/optionality of a field), please follow these rules:
  - If the API is unusable without the change -> every supported version
  - If the API is usable, but fix is on the response side -> every supported version
  - If the API is usable, but fix is on the request side -> no backport, unless the API is _partially_ usable and the fix unlocks a missing feature that has no workaround

Happy coding!

-->

👋 Hey, I'm not sure if it's a best practice to add the client types after the FF. If I mislabelled it please let me know. The connectors API actions are introduced in `8.12`.

### Changes 

- Copy endpoints from es [rest-api-spec](https://github.com/elastic/elasticsearch/tree/8.12/rest-api-spec/src/main/resources/rest-api-spec/api) 
- Copy and adapt types from [Kibana ts definition](https://github.com/elastic/kibana/blob/main/packages/kbn-search-connectors/types/connectors.ts)
- Added types for
  - connector.put
  - connector.post
  - connector.delete
  - connector.get
  - connector.list
  - connector.check_in
  - connector.update_error
  - connector.update_scheduling
  - connector.update_filtering
  - connector.update_configuration
  - connector.update_pipeline
  - connector.last_sync <- it should be `update_last_sync`, we should followup to fix this in 8.13 on ES side

jedrazb commented on output/schema/validation-errors.json:
```diff
@@ -482,6 +482,12 @@
+    "connector.delete": {
+      "request": [],
+      "response": [
+        "response definition connector.delete:Response / body / instance_of - Non-leaf type cannot be used here: '_types:AcknowledgedResponseBase'"
```
This issue is common for other endpoints that deal with `delete` action - should I change something here?

Anaethelion replied:
No, you can leave it as is.

Anaethelion commented on specification/connector/_types/Connector.ts:
```diff
@@ -0,0 +1,241 @@
+  basic?: FeatureEnabled
+}
+
+export interface ConnectorFeatures {
+  [FeatureName.DOCUMENT_LEVEL_SECURITY]?: FeatureEnabled
+  [FeatureName.FILTERING_ADVANCED_CONFIG]?: boolean
+  [FeatureName.FILTERING_RULES]?: boolean
+  [FeatureName.INCREMENTAL_SYNC]?: FeatureEnabled
+  [FeatureName.SYNC_RULES]?: SyncRulesFeature
+}
```
```suggestion
export interface ConnectorFeatures {
  document_level_security?: FeatureEnabled
  filtering_advanced_config?: boolean
  filtering_rules?: boolean
  incremental_sync?: FeatureEnabled
  sync_rules?: SyncRulesFeature
}
```

That mechanism isn't implemented in the compiler, this results in an object with the properties called `[FeatureName.DOCUMENT_LEVEL_SECURITY]` which isn't what you wanted I presume.

These properties should be name directly from their counterparts in `FeatureName` and the enum removed.

Anaethelion commented on specification/connector/_types/Connector.ts:
```diff
@@ -0,0 +1,241 @@
+  run_ml_inference: boolean
+}
+
+type FilteringPolicy = 'exclude' | 'include'
+
+type FilteringRuleRule =
+  | 'contains'
+  | 'ends_with'
+  | 'equals'
+  | '>'
+  | '<'
+  | 'regex'
+  | 'starts_with'
```
`FilteringPolicy` and `FilteringRuleRule` should be converted to enums.

Anaethelion commented on specification/connector/_types/Connector.ts:
```diff
@@ -0,0 +1,241 @@
+  BOOLEAN = 'bool'
+}
+
+interface ConnectorConfigCategoryProperties {
```
This seems redundant with `ConnectorConfigProperties` and I can't find this in the server sources. If its not used, this should be removed and the `ConnectorConfiguration` simplified.

jedrazb replied:
Agreed, we can get rid of it (also checked in Kibana source).

Anaethelion commented on specification/connector/_types/Connector.ts:
```diff
@@ -0,0 +1,241 @@
+  draft: FilteringRules
+}
+
+enum FeatureName {
+  FILTERING_ADVANCED_CONFIG = 'filtering_advanced_config',
+  FILTERING_RULES = 'filtering_rules',
+  DOCUMENT_LEVEL_SECURITY = 'document_level_security',
+  INCREMENTAL_SYNC = 'incremental_sync',
+  SYNC_RULES = 'sync_rules'
+}
```
```suggestion
```

Anaethelion commented on specification/connector/_types/Connector.ts:
```diff
@@ -0,0 +1,241 @@
+  [FeatureName.SYNC_RULES]?: SyncRulesFeature
+}
+
+export interface SchedulingConfiguraton {
```
```suggestion
export interface SchedulingConfiguration {
```

Nit: typo, should also be updated wherever referenced.

Anaethelion reviewed (changes requested):
Left a few comments.

I've also taken the liberty to add the links to the documentation within the endpoints, feel free to edit/update if I made any mistake.

jedrazb commented:
Hey @Anaethelion thank you for review and adding the documentation links 👍 . I addressed your feedback and also updated the json spec files with correct doc links. 

There are couple of more endpoints to be added (related to `_connector/_sync _job` API). Given the release timeline is it still ok to keep the `backport 8.12` label?

jedrazb commented:
Hey @swallez 👋  since Anaethelion is OOO this week, I was suggested to add you as a reviewer to this PR. Could you take a look?

jedrazb commented on specification/connector/_types/Connector.ts:
```diff
@@ -0,0 +1,231 @@
+}
+
+enum DisplayType {
+  TEXTBOX = 'textbox',
```
Thank you for pointing this out 👍 I'm guilty of just copy-pasting most of the types from defs in Kibana

jedrazb commented on specification/connector/_types/Connector.ts:
```diff
@@ -0,0 +1,231 @@
+}
+
+interface CustomScheduling {
+  configuration_overrides: Dictionary<string, UserDefinedValue>
```
Yeah, my bad, it was defined like that [in Kibana](https://github.com/elastic/kibana/blob/main/packages/kbn-search-connectors/types/connectors.ts#L80), I fill fix this here and file a follow-up ticket to handle it in Kibana as well.

jedrazb commented on specification/connector/_types/Connector.ts:
```diff
@@ -0,0 +1,231 @@
+
+export type ConnectorConfiguration = Dictionary<
+  string,
+  ConnectorConfigProperties | null
```
Again, I'm guilty of copy-pasting from (probably) slightly outdated types from [Kibana](https://github.com/elastic/kibana/blob/main/packages/kbn-search-connectors/types/connectors.ts#L65). According to my understanding `| null` here is indeed not needed

jedrazb commented on specification/connector/_types/Connector.ts:
```diff
@@ -0,0 +1,231 @@
+  display: DisplayType
+  label: string
+  options: SelectOption[]
+  order?: number | null
```
I revised the types and I think for majority of cases we can safely mark fields as optional (instead of making them nullable). The only exception are errors that are (string | null), I can think of some situations where we want to mark error as null (when connector gets back to a healthy state)

swallez replied:
Ok. So we have to implement https://github.com/elastic/elasticsearch-specification/issues/2049 that clearly distinguishes "null as optional" and "null as an actual value" in the spec. Adding that on my immediate todo list.

jedrazb replied:
Does it imply that this PR needs to wait until #2049 is implemented? We are not in a rush, just want to clarify

swallez replied:
Yes, because currently `null` is considered as the equivalent of "missing" in languages where objects are represented as data structures and not dictionaries (i.e. all strictly typed languages). #2049 is meant to capture "null is an actual value" in the API spec so that we can generate code that correctly handles this.

jedrazb commented on specification/connector/_types/Connector.ts:
```diff
@@ -0,0 +1,231 @@
+}
+
+interface Validation {
+  constraint: string | number
```
It's limited to strings and numbers. It's represented as an `Object` type in ES, though the [parser](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/configuration/ConfigurationValidation.java#L65) validates that it's either of numeric or string type

swallez replied:
Oh, as I understand, `type` is an [enumeration of the possible constraints](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/configuration/ConfigurationValidationType.java) that precisely defines what should be in the `constraint` field.

So this can be represented in a fully-typed way as follows using an union defining "[internally tagged variants](https://github.com/elastic/elasticsearch-specification/blob/main/docs/modeling-guide.md#internal)":

```ts
/** @variants internal tag='type' */
export type Validation = LessThanValidation | GreaterThanValidation | ListTypeValidation | IncludedInValidation | RegexValidation | UnsetValidation

export interface LessThanValidation {
  type: 'less_than'
  constraint: double // we want to avoid 'number'
}

export interface GreaterThanValidation {
  type : 'greater_than'
  constraint: double
}

export interface ListTypeValidation {
  type: 'list_type'
  constraint: string
}

export interface IncludedInValidation {
  type: 'included_in'
  constraint: string // is it really a string?
}

export interface RegexValidation {
  type: 'regex'
  constraint: string
}

export interface UnsetValidation {
  type: 'unset'
  // no constraint ?
}
```

jedrazb replied:
Ack, thank you for explanation! Will include this into the type definitions

jedrazb commented:
CI steps are breaking in the `setup job` phase with 
```
Error: Internal server error occurred while resolving "actions/checkout@v[2](https://github.com/elastic/elasticsearch-specification/actions/runs/7461636399/job/20302186145?pr=2367#step:1:2)"
```

swallez commented on specification/connector/list/ConnectorListRequest.ts:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * A comma-separated list of connector index names to fetch connector documents for
+     */
+    index_name?: Fields
```
`Fields` is meant to hold a list of index field names. If these are ES indices, then `Indices` should be used.

swallez commented on specification/connector/list/ConnectorListRequest.ts:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * A comma-separated list of connector names to fetch connector documents for
+     */
+    connector_name?: Fields
```
Use `Names` instead

swallez commented on specification/connector/list/ConnectorListRequest.ts:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * A comma-separated list of connector service types to fetch connector documents for
+     */
+    service_type?: Fields
```
Are service type identifiers? Then use `Ids`, or otherwise `Names`.

jedrazb replied:
They are arbitrary strings, so it is `Names` for our case

swallez reviewed (changes requested):
Looking good! There are a few minor naming issues left, and we'll be good to go!

l-trotta commented on specification/connector/_types/Connector.ts:
```diff
@@ -0,0 +1,267 @@
+  configuration: ConnectorConfiguration
+  custom_scheduling: ConnectorCustomScheduling
+  description?: string
+  error: WithNullValue<string>
```
error should be nullable, [server code](https://github.com/elastic/elasticsearch/blob/92f7e078071fba021f9784c98a7cea6e76c5710d/x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java#L82)

pquentin commented:
"Validate APIs / build" was failing because of me, I fixed it and retried the build.

jedrazb commented on specification/connector/_types/Connector.ts:
```diff
@@ -0,0 +1,267 @@
+  configuration: ConnectorConfiguration
+  custom_scheduling: ConnectorCustomScheduling
+  description?: string
+  error: WithNullValue<string>
```
Yes, I do agree with you here. Since the `Connector` denotes a response type. In a response, a missing (optional) field does indicate that the field is in fact null. 

I will adapt this, but I the meaningful is needed for requests definitions to set certain fields to null.

swallez reviewed (approved):
LGTM! 🎉

flobernd commented on specification/connector/update_index_name/ConnectorUpdateIndexNameRequest.ts:
```diff
@@ -0,0 +1,43 @@
+   * The connector index name
+   */
+  body: {
+    index_name: WithNullValue<string>
```
@swallez @Anaethelion I'm late to the party, but shouldn't we use the `IndexName` alias instead of `string` at these places?

Like:

```ts
index_name: WithNullValue<IndexName>
```
